### PR TITLE
feat: worker on the GPU

### PR DESCRIPTION
**Pull request by salma-aneo**

### CODE REVIEW

File: python/hello-world-gpu/worker/worker.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    logger = ClefLogger.getLogger("ArmoniKWorker")
+    logger.info("Handeling the Task")
+
+    if not cuda.is_available():
+        raise RuntimeError(
+            "CUDA is not available. Please check your GPU and driver installation."
+        )
+    logger.info("CUDA is available.")
+    logger.info(f"Device name: {cuda.get_current_device().name}")
```
Comment (ngruelaneo) [inline python/hello-world-gpu/worker/worker.py:41]: pas de fstring en logger

File: python/hello-world-gpu/worker/worker.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    # Launch kernel
+    vector_add[blocks_per_grid, threads_per_block](a_device, b_device, c_device)
+    cuda.synchronize()  # Wait for the kernel to finish
+    end = time.time()
+
+    # Copy the result back to the host (CPU)
+    c_host = c_device.copy_to_host()
+
+    # Print and verify results
+    logger.info("GPU execution time:" + str(end - start) + "seconds")
```
Comment (ngruelaneo) [inline python/hello-world-gpu/worker/worker.py:84]: cela donne quoi avec le logger ClefLogger?

File: python/hello-world-gpu/readme.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+The latest NVIDIA Windows GPU Driver will fully support WSL 2. With CUDA support in the driver, existing applications (compiled elsewhere on a Linux system for the same target GPU) can run unmodified within the WSL environment.
+
+To compile new CUDA applications, a CUDA Toolkit for Linux x86 is needed.
+
+Once a Windows NVIDIA GPU driver is installed on the system, CUDA becomes available within WSL 2. The CUDA driver installed on Windows host will be stubbed inside the WSL 2 as libcuda.so, therefore users must not install any NVIDIA GPU Linux driver within WSL 2. One has to be very careful here as the default CUDA Toolkit comes packaged with a driver, and it is easy to overwrite the WSL 2 NVIDIA driver with the default installation. We recommend developers to use a separate CUDA Toolkit for WSL 2 (Ubuntu) available from the CUDA Toolkit Downloads page to avoid this overwriting. This WSL-Ubuntu CUDA toolkit installer will not overwrite the NVIDIA driver that was already mapped into the WSL 2 environment.
+
+First, remove the old GPG key:
+    
+    sudo apt-key del 7fa2af80
```
Comment (ngruelaneo) [inline python/hello-world-gpu/readme.md:48]: use the bash environment to higlight this part.

File: python/hello-world-gpu/readme.md [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+Restart Docker Daemon:
+
+After saving the changes, restart the Docker daemon to apply the new configuration.
+
+```bash
+sudo systemctl restart docker
+```
+
+### 6. NVIDIA device plugin for Kubernetes
+
```
Comment (ngruelaneo) [inline python/hello-world-gpu/readme.md:225]: This part should be in the terraform deployment I think.

File: python/hello-world-gpu/readme.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    pip install -r client-requirements.txt
+    ```
+
+## Usage
+
+Run the Client with the name of the partition:
+
+```bash
+python client.py --partition helloworldgpu
+```
```
Comment (ngruelaneo) [inline python/hello-world-gpu/readme.md:269]: add the result expected

File: python/hello-world-gpu/common/common.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        This method converts the numpy array to a list, encodes it as JSON,
+        and then encodes it to bytes.
+
+        Returns:
+            bytes: The serialized numpy array as a byte array.
+        """
+        array_list = self.array.tolist()  # Convert numpy array to list
+        array_json = json.dumps(array_list)  # Serialize list to JSON string
+        return array_json.encode("utf-8")  # Encode JSON string to bytes
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/common/common.py:72]:
```suggestion
        return self.array.to_bytes()
```

File: python/hello-world-gpu/common/common.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        Args:
+            payload (bytes): The serialized data as bytes.
+
+        Returns:
+            NumpyArraySerializer: An instance of NumpyArraySerializer created from the serialized data.
+        """
+        array_json = payload.decode("utf-8")  # Decode bytes to JSON string
+        array_list = json.loads(array_json)  # Deserialize JSON string to list
+        array = np.array(array_list, dtype=np.float32)  # Convert list to numpy array
+        return cls(array)
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/common/common.py:91]:
```suggestion
        array = np.frombuffer(array_list, dtype=np.float32)
        return cls(array)
```

File: python/hello-world-gpu/worker/worker.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    Processes a task by summing the two vectors and sending the result.
+
+    Args:
+        task_handler: The handler for the current task.
+
+    Returns:
+        Output: The result of the task processing.
+    """
+    logger = ClefLogger.getLogger("ArmoniKWorker")
+    logger.info("Handeling the Task")
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/worker/worker.py:34]:
```suggestion
    logger.info("Handling the Task")
```

**Review by dbrasseur-aneo — CHANGES_REQUESTED** [review]
Just details

File: python/hello-world-gpu/readme.md (hunk 4 above)
Comment (lemaitre-aneo) [reply python/hello-world-gpu/readme.md:234]: This is related to the deployment of Kubernetes itself, not ArmoniK. So it makes sense that is it a separated step.

Comment (salma-aneo) [reply python/hello-world-gpu/readme.md:234]: if deploying in local, one should add the plugin to the cluster. In the cloud this step ain't needed.

File: python/hello-world-gpu/client/client.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
+# Add the common directory to the system path
+common_path = Path(__file__).resolve().parent.parent / "common"
+sys.path.append(str(common_path))
+
+from common import NameIdDict, NumpyArraySerializer
```
[the remark this answers is not in the document]
Comment (dbrasseur-aneo) [reply python/hello-world-gpu/client/client.py:23]: Honestly client common and worker should not be in their own folders, which would turn it into:
```suggestion
from .common import NameIdDict, NumpyArraySerializer
```

File: python/hello-world-gpu/worker/worker.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+from pathlib import Path
+import grpc
+import time
+
+# Add the common directory to the system path
+common_path = Path(__file__).resolve().parent.parent / "common"
+sys.path.append(str(common_path))
+
+
+from common import NameIdDict, NumpyArraySerializer
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/worker/worker.py:17]: Same as previous comment

File: python/hello-world-gpu/client.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                "Result ready",
+                extra={"context": {"resultId": output_id, "data": final_result}},
+            )
+        except Exception as e:
+            logger.error("An error occured", extra={"context": {"error": e.details}})
+
+    logger.info("End Connection!")
+
+
+def main(args: List[str]) -> None:
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/client.py:147]: You don't need to pass the arguments as a list, as just calling parse_args with no args already take the arguments of the command line

File: python/hello-world-gpu/client.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
+import argparse
+import logging
+from typing import List
+import grpc
+import grpc.aio
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/client.py:5]:
```suggestion
```
You don't use features of async grpc

File: python/hello-world-gpu/client.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+def run(endpoint: str, partition: str, size: int, seed: int) -> None:
+    """
+    Connects to the ArmoniK control plane via a gRPC channel and performs a series of tasks.
+
+    Args:
+        endpoint: The endpoint for the connection to ArmoniK control plane.
+        partition: The name of the partition to which tasks are submitted.
+        size: the size of the two vectors.
+
```
Comment (dbrasseur-aneo) [inline python/hello-world-gpu/client.py:28]: Missing the seed parameter documentation

**Review by dbrasseur-aneo — COMMENTED** [review]
No more softlink fix needed ?

**Review by dbrasseur-aneo — APPROVED** [review]

**Pull request merged**